boot/assets.go

```diff
@@ -335,16 +367,171 @@ type TrustedAssetsUpdateObserver struct{}
 //
 // Implements gadget.ContentUpdateObserver.
 func (o *TrustedAssetsUpdateObserver) Observe(op gadget.ContentOperation, affectedStruct *gadget.LaidOutStructure, root, realSource, relativeTarget string) (bool, error) {
```

Comment:
While working on mark boot successful code which will leave only the hashes of the assets we booted with I came up with a scenario that will cause trouble and discussed it with @pedronis. The scenario is as follows:

- initial state: grub: [A]
- update runs -> grub: [A, B] (B is the new revision)
 - B gets written to disk
- spontaneous reboot
- marksuccessful runs -> grub: [B] (only hashes that we booted with)
  - marksuccessful reseals with grub: [B]
- update picks up
- backup pass runs -> grub: [B] (since B is already tracked in map, and current code does not know about A)
- BeforeWrite gets called, we reseal with [B] only
- something in update errors out and rollback is triggered
- rollback runs, restores A
  - FAIL: observe is called with "" source, and relative path pointing to restored A, check of expected hash (which would be B) and on disk hash (A) fails
  - FAIL: if we reboot at this point, the system will not boot, because key is sealed to B only

I will be addressing this and passing the paths to the old asset and the new one, so that the observer of an update, can ensure that we get grub: [A, B] in modeenv. 

This being said, the PR should land, and the fix should come on top to keep the things clean. Nonetheless, it needs to land before landing the code in mark successful.